File: site/src/lib/first-follow.ts
// FIRST / FOLLOW / nullable, computed client-side from the grammar's own
// productions. The engine does not expose these sets yet (playground-spec §5),
// but they are a pure function of the productions — so we derive them in TS
// rather than fake them. Reuses the diagram parser's view of a rule (its
// alternatives as sequences of terminal/nonterminal symbols), so this and the
// railroad diagrams read the grammar identically.
import type { Production, DiaSym } from "../../../bootstrap/railroad.ts";

export interface FirstFollow {
  /** Nonterminal -> its FIRST set (terminal labels). */
  first: Record<string, string[]>;
  /** Nonterminal -> its FOLLOW set (terminal labels; `$` = end of input). */
  follow: Record<string, string[]>;
  /** The nonterminals that can derive the empty string. */
  nullable: string[];
}

const END = "$";

/** Fixed-point FIRST/FOLLOW over a flat (epsilon-only-via-empty-alt) grammar.
 * The first production's LHS is the start symbol (gets `$` in its FOLLOW). */
export function computeFirstFollow(prods: Production[]): FirstFollow {
  const nts = new Set(prods.map((p) => p.name));
  const isNt = (s: DiaSym) => !s.term && nts.has(s.label);

  // --- nullable (fixed point) ---
  const nullable = new Set<string>();
  for (let changed = true; changed;) {
    changed = false;
    for (const p of prods) {
      if (nullable.has(p.name)) continue;
      const canBeEmpty = p.alts.some((alt) =>
        alt.every((s) => isNt(s) && nullable.has(s.label)),
      );
      if (canBeEmpty) {
        nullable.add(p.name);
        changed = true;
      }
    }
  }

  const first: Record<string, Set<string>> = {};
  const follow: Record<string, Set<string>> = {};
  for (const n of nts) {
    first[n] = new Set();
    follow[n] = new Set();
  }

  // FIRST of a symbol: a terminal is itself; a nonterminal, its FIRST set.
  const firstOf = (s: DiaSym): Set<string> =>
    isNt(s) ? first[s.label]! : new Set([s.label]);

  // --- FIRST (fixed point) ---
  for (let changed = true; changed;) {
    changed = false;
    for (const p of prods) {
      for (const alt of p.alts) {
        let allNullable = true;
        for (const sym of alt) {
          for (const t of firstOf(sym)) {
            if (!first[p.name]!.has(t)) {
              first[p.name]!.add(t);
              changed = true;
            }
          }
          if (!(isNt(sym) && nullable.has(sym.label))) {
            allNullable = false;
            break;
          }
        }
        // An all-nullable (or empty) alternative contributes nothing extra to
        // FIRST here; nullability is tracked separately.
        void allNullable;
      }
    }
  }

  // --- FOLLOW (fixed point) ---
  if (prods[0]) follow[prods[0].name]!.add(END);
  for (let changed = true; changed;) {
    changed = false;
    for (const p of prods) {
      for (const alt of p.alts) {
        for (let i = 0; i < alt.length; i++) {
          const sym = alt[i]!;
          if (!isNt(sym)) continue;
          const target = follow[sym.label]!;
          // FIRST of the remainder β; if β is all-nullable, add FOLLOW(LHS).
          let betaNullable = true;
          for (let j = i + 1; j < alt.length; j++) {
            const b = alt[j]!;
            for (const t of firstOf(b)) {
              if (!target.has(t)) {
                target.add(t);
                changed = true;
              }
            }
            if (!(isNt(b) && nullable.has(b.label))) {
              betaNullable = false;
              break;
            }
          }
          if (betaNullable) {
            for (const t of follow[p.name]!) {
              if (!target.has(t)) {
                target.add(t);
                changed = true;
              }
            }
          }
        }
      }
    }
  }

  const sort = (s: Set<string>) =>
    [...s].sort((a, b) =>
      a === END ? 1 : b === END ? -1 : a.localeCompare(b),
    );
  const out: FirstFollow = { first: {}, follow: {}, nullable: [...nullable] };
  for (const n of nts) {
    out.first[n] = sort(first[n]!);
    out.follow[n] = sort(follow[n]!);
  }
  return out;
}

File: site-glue/js/src/main/scala/gramaire/site/FirstFollow.scala
package gramaire.site

import scala.scalajs.js
import scala.scalajs.js.JSConverters.*
import scala.scalajs.js.annotation.JSExportTopLevel
import gramaire.Railroad.{DiaSym, Production}

// FIRST / FOLLOW / nullable, computed client-side from the grammar's own
// productions. The engine does not expose these sets (playground-spec §5),
// but they are a pure function of the productions — so they're derived here
// rather than faked. Reuses the diagram parser's view of a rule (its
// alternatives as sequences of terminal/nonterminal symbols), so this and the
// railroad diagrams read the grammar identically.
//
// Deliberately its own algorithm, not a reuse of `gramaire.Analyze` (the
// `bootstrap/analyze.ts` port used by `gramaire fmt`'s table regeneration):
// that one assumes epsilon-free, already-desugared grammars, while this one
// tracks nullability explicitly — the Lab's editor text is often a half-typed,
// not-yet-valid grammar, so it can't make that assumption.
// Ported from site/src/lib/first-follow.ts.
object FirstFollow:

  private val END = "$"

  @js.native
  trait Result extends js.Object:
    val first: js.Dictionary[js.Array[String]]
    val follow: js.Dictionary[js.Array[String]]
    val nullable: js.Array[String]

  /** Fixed-point FIRST/FOLLOW over a flat (epsilon-only-via-empty-alt) grammar. The first
    * production's LHS is the start symbol (gets `$` in its FOLLOW).
    */
  @JSExportTopLevel("computeFirstFollow")
  def computeFirstFollow(prods: js.Array[Production]): Result =
    val ps = prods.toVector
    val nts = ps.map(_.name).toSet
    def isNt(s: DiaSym): Boolean = !s.term && nts.contains(s.label)

    // --- nullable (fixed point) ---
    var nullable = Set.empty[String]
    var changed = true
    while changed do
      changed = false
      for p <- ps if !nullable.contains(p.name) do
        val canBeEmpty = p.alts.exists(_.forall(s => isNt(s) && nullable.contains(s.label)))
        if canBeEmpty then
          nullable += p.name
          changed = true

    var first: Map[String, Set[String]] = nts.map(_ -> Set.empty[String]).toMap
    var follow: Map[String, Set[String]] = nts.map(_ -> Set.empty[String]).toMap

    // FIRST of a symbol: a terminal is itself; a nonterminal, its FIRST set.
    def firstOf(s: DiaSym): Set[String] = if isNt(s) then first(s.label) else Set(s.label)

    // --- FIRST (fixed point) ---
    changed = true
    while changed do
      changed = false
      for p <- ps; alt <- p.alts do
        // FIRST(lhs) gains FIRST of each symbol up to and including the
        // first non-nullable one — a symbol after that point is never at
        // the leftmost position of a derivation from `p`, so it must NOT
        // contribute to FIRST(lhs) (unlike the whole-alternative scan this
        // loop must stop, not just skip nullability bookkeeping).
        var k = 0
        var stopAlt = false
        while k < alt.length && !stopAlt do
          val sym = alt(k)
          val before = first(p.name)
          val after = before ++ firstOf(sym)
          if after != before then
            first = first.updated(p.name, after)
            changed = true
          if !(isNt(sym) && nullable.contains(sym.label)) then stopAlt = true
          k += 1

    // --- FOLLOW (fixed point) ---
    ps.headOption.foreach(p0 => follow = follow.updated(p0.name, follow(p0.name) + END))
    changed = true
    while changed do
      changed = false
      for p <- ps; alt <- p.alts do
        for i <- alt.indices do
          val sym = alt(i)
          if isNt(sym) then
            val target = follow(sym.label)
            var betaNullable = true
            var updated = target
            var j = i + 1
            var stop = false
            while j < alt.length && !stop do
              val b = alt(j)
              updated = updated ++ firstOf(b)
              if !(isNt(b) && nullable.contains(b.label)) then
                betaNullable = false
                stop = true
              j += 1
            val finalSet = if betaNullable then updated ++ follow(p.name) else updated
            if finalSet != target then
              follow = follow.updated(sym.label, finalSet)
              changed = true

    // `localeCompare` (not codepoint order) to byte-match the TS reference's
    // `a.localeCompare(b)` — ICU collation orders punctuation differently
    // from raw codepoints, so this can't be a plain Scala string comparison.
    def localeCompare(a: String, b: String): Int =
      a.asInstanceOf[js.Dynamic].localeCompare(b).asInstanceOf[Double].toInt

    def sort(s: Set[String]): js.Array[String] =
      s.toVector
        .sortWith((a, b) =>
          if a == END then false else if b == END then true else localeCompare(a, b) < 0
        )
        .toJSArray

    val out = js.Dynamic.literal(
      first = js.Dictionary(nts.toVector.map(n => n -> sort(first(n)))*),
      follow = js.Dictionary(nts.toVector.map(n => n -> sort(follow(n)))*),
      nullable = nullable.toVector.toJSArray
    )
    out.asInstanceOf[Result]
